From d2758a0e6fb61cc232a7262cd3c97165c2800caf Mon Sep 17 00:00:00 2001
From: Chris Hofer <christian.hofer@codasip.com>
Date: Thu, 26 Jun 2025 13:47:44 +0200
Subject: [PATCH] libcap: Make raw container_of calculation platform
 independent

libcap uses for the struct:
	struct _cap_alloc_s {
	    __u32 magic;
	    __u32 size;
	    union {
		...
    	    } u;
	};

the expression:
	void *base = (void *) (-2 + (__u32 *) union_ptr);

to obtain a pointer to the struct from a pointer to the union. So it
subtracts 8 bytes from the union pointer. As the union can hold a
pointer, this calculation fails due to the union alignment on platforms
where pointers have larger alignment constraints than 8 bytes.

Fix this by calculating and subtracting the actual number of 32-bit
unsigned integers from the union_ptr based on a compile time byte
offset. The byte offset is always a multiple of the byte offset as the
union member has stricter alignment constraints than the 32-bit members.
Nevertheless, statically assert it.

This solution was suggested by Stuart Menefy and Florian Schmaus.

Upstream-Status: Pending
Signed-off-by: Chris Hofer <christian.hofer@codasip.com>
---
 libcap/cap_alloc.c | 6 +++++-
 libcap/libcap.h    | 5 ++++-
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/libcap/cap_alloc.c b/libcap/cap_alloc.c
index 504abd2..7802aaa 100644
--- a/libcap/cap_alloc.c
+++ b/libcap/cap_alloc.c
@@ -5,6 +5,8 @@
  * capability sets as specified by POSIX.1e (formerlly, POSIX 6).
  */
 
+#include <stddef.h>
+#include <assert.h>
 #include "libcap.h"
 
 /*
@@ -48,6 +50,8 @@ struct _cap_alloc_s {
 	struct cap_launch_s launcher;
     } u;
 };
+const __u32 offset_magic = offsetof(struct _cap_alloc_s, u);
+static_assert((offsetof(struct _cap_alloc_s, u) % sizeof(__u32)) == 0);
 
 /*
  * Obtain a blank set of capabilities
@@ -266,7 +270,7 @@ int cap_free(void *data_p)
 	return -1;
     }
 
-    void *base = (void *) (-2 + (__u32 *) data_p);
+    void *base = (void *) (((__u32 *) data_p) - (offset_magic / sizeof(__u32)));
     struct _cap_alloc_s *data = base;
     switch (data->magic) {
     case CAP_T_MAGIC:
diff --git a/libcap/libcap.h b/libcap/libcap.h
index f4a72fe..cddfbbc 100644
--- a/libcap/libcap.h
+++ b/libcap/libcap.h
@@ -153,7 +153,10 @@ struct _cap_struct {
 /* launcher magic for cap_free */
 #define CAP_LAUNCH_MAGIC 0xCA91AC
 
-#define magic_of(x)           ((x) ? *(-2 + (const __u32 *) x) : 0)
+extern const __u32 offset_magic;
+#define magic_of(x)           ((x) ? \
+                              *(((const __u32 *) x) - (offset_magic / sizeof(__u32))) : \
+                              0)
 #define good_cap_t(x)         (CAP_T_MAGIC   == magic_of(x))
 #define good_cap_iab_t(x)     (CAP_IAB_MAGIC == magic_of(x))
 #define good_cap_launch_t(x)  (CAP_LAUNCH_MAGIC == magic_of(x))
-- 
2.43.5

